### DISCUSSION TITLE
Small part locations

### DISCUSSION THREAD

**Discussion by smurfix** [inventree/InvenTree#8404] [Q&A]
Hello,
part of my lab organization consists of magazine drawers, mostly 3d-printed, with varying compartment sizes so that I can use the smallest box that can hold e.g. all my 4-pin 2mm JST 90° connectors, or the 1/4W 0604 12k resistors, thereby saving a lot of shelf space.

However. This has implications that currently don't seem to fit InvenTree.

* All those locations (100 drawers, each with anywhere from 3 to 50 compartments …) blow up the Location pop-up menu. Can it be changed to behave hierarchically? Or maybe there's another way to do this?

I can only assume that people who have reels with parts instead of little bins have the same problem; the reel of 0604 1.2k 1/4W resistors might be in rack 17 level 5 position 10 (of 60) …?

* Part of my location management is to find free space for things (and to not misplace stuff). I need some way to do "find a location of size X with no part in it".

* A "this location can only contain one distinct stock item" flag would make sense.

**Comment by SchrodingersGat** [comment]
@smurfix some interesting ideas here!

> Can it be changed to behave hierarchically? Or maybe there's another way to do this?

Locations can be arranged in a tree / hierarchy structure already - is this what you are asking?

> I need some way to do "find a location of size X with no part in it".

A plugin to achieve this might be a good solution here? 

> A "this location can only contain one distinct stock item" flag would make sense.

Again you could potentially achieve this with a plugin, I am reluctant to add single-purpose flags like this to the general codebase

**Reply by wolflu05** [reply]
> I need some way to do "find a location of size X with no part in it".

That's exactly what I was planning to add in the future with the already existing location types.

**Reply by smurfix** [reply]
> Locations can be arranged in a tree / hierarchy structure already - is this what you are asking?

Well, the problem is that the menu for picking a location doesn't have those little expansion buttons. When the location list gets ten thousand long things get unwieldy. And slow.

That, a filter for empty locations, and a filter for location type (like "only show 2×3×1cm drawer compartments" except that it shouldn't squash the location menu's hierarchy) would go a **long** way towards fixing my problem.

We can think about the fact that these drawers are movable and thus might want to be re-parented some other time. Ideally "where is it" boils down to "Lab 2, magazine 3, row 4 [column 2], drawer #312, row 8 col 3: capacitor 10µF 0805 30V" but of course that's way too complex, esp. given that the drawers themselves can be relocated.

**Reply by wolflu05** [reply]
Yes, this is currently also not optimal. I also have in mind to completely change the input component for tree fields by default with a tree picker, similar to the one in cui you can open with the little arrow on the right, but much better performancewise. But unfortunately I had no time yet to implement it.

(FYI, if you have lots of drawers, you may find [this plugin](https://github.com/wolflu05/inventree-bulk-plugin) helpful, to automatically create all of those locations)
